File: tests/test_graph.py
from data_structures.graph import Graph


def build_sample_graph():
    g = Graph(directed=True)
    g.add_edge("A", "B", 1)
    g.add_edge("A", "C", 4)
    g.add_edge("B", "C", 2)
    g.add_edge("B", "D", 5)
    g.add_edge("C", "D", 1)
    return g


def test_bfs_order():
    g = build_sample_graph()
    order = g.bfs("A")
    assert order[0] == "A"
    assert set(order) == {"A", "B", "C", "D"}


def test_dfs_reaches_all_connected_vertices():
    g = build_sample_graph()
    order = g.dfs("A")
    assert order[0] == "A"
    assert set(order) == {"A", "B", "C", "D"}


def test_bfs_dfs_on_unknown_start_returns_empty():
    g = build_sample_graph()
    assert g.bfs("Z") == []
    assert g.dfs("Z") == []


def test_dijkstra_shortest_distances():
    g = build_sample_graph()
    distances = g.dijkstra("A")
    assert distances["A"] == 0
    assert distances["B"] == 1
    # A->C direct is 4, but A->B->C is 1+2=3, which is shorter.
    assert distances["C"] == 3
    # A->B->D is 1+5=6; A->B->C->D is 1+2+1=4, which is shorter.
    assert distances["D"] == 4


def test_dijkstra_unreachable_vertex_is_infinity():
    g = build_sample_graph()
    g.add_vertex("Isolated")
    distances = g.dijkstra("A")
    assert distances["Isolated"] == float("inf")


def test_dijkstra_rejects_negative_weights():
    g = Graph(directed=True)
    g.add_edge("A", "B", -1)
    import pytest

    with pytest.raises(ValueError):
        g.dijkstra("A")


def test_undirected_graph_edge_is_symmetric():
    g = Graph(directed=False)
    g.add_edge("A", "B", 1)
    assert ("B", 1) in g.neighbors("A")
    assert ("A", 1) in g.neighbors("B")

File: tests/test_graph.cpp
#include <limits>

#include "../include/data_structures/graph.h"
#include "../include/test_framework.h"

TEST_CASE("graph: bfs visits in breadth-first order") {
    Graph g(6);
    g.add_edge(0, 1);
    g.add_edge(0, 2);
    g.add_edge(1, 3);
    g.add_edge(2, 4);
    g.add_edge(3, 5);
    auto order = g.bfs(0);
    std::vector<int> expected = {0, 1, 2, 3, 4, 5};
    CHECK(order == expected);
}

TEST_CASE("graph: dfs visits depth-first before backtracking") {
    Graph g(5);
    g.add_edge(0, 1);
    g.add_edge(0, 2);
    g.add_edge(1, 3);
    g.add_edge(1, 4);
    auto order = g.dfs(0);
    std::vector<int> expected = {0, 1, 3, 4, 2};
    CHECK(order == expected);
}

TEST_CASE("graph: bfs/dfs on a single isolated vertex") {
    Graph g(1);
    CHECK(g.bfs(0) == std::vector<int>{0});
    CHECK(g.dfs(0) == std::vector<int>{0});
}

TEST_CASE("graph: dijkstra finds shortest paths on a weighted graph") {
    Graph g(5);
    g.add_edge(0, 1, 4);
    g.add_edge(0, 2, 1);
    g.add_edge(2, 1, 2);
    g.add_edge(1, 3, 1);
    g.add_edge(2, 3, 5);
    g.add_edge(3, 4, 3);
    auto dist = g.dijkstra(0);
    // shortest 0->1 is via 2 (1+2=3), not the direct edge (4)
    CHECK(dist[1] == 3);
    CHECK(dist[2] == 1);
    CHECK(dist[3] == 4);
    CHECK(dist[4] == 7);
}

TEST_CASE("graph: dijkstra leaves unreachable vertices at INT_MAX") {
    Graph g(3);
    g.add_edge(0, 1, 1);
    // vertex 2 has no edges at all
    auto dist = g.dijkstra(0);
    CHECK(dist[0] == 0);
    CHECK(dist[1] == 1);
    CHECK(dist[2] == std::numeric_limits<int>::max());
}

TEST_CASE("graph: directed edges only traverse one way") {
    Graph g(3);
    g.add_edge(0, 1, 1, /*directed=*/true);
    auto dist_from_0 = g.dijkstra(0);
    auto dist_from_1 = g.dijkstra(1);
    CHECK(dist_from_0[1] == 1);
    CHECK(dist_from_1[0] == std::numeric_limits<int>::max());
}
